Adding basic end conditions to game

patrickkoopmans opened the pull request:
This pull request to merge a small collaborative feature between WDL and Game State to add an end condition to games. 

As specified by WDL, the end condition will consist of a final_room specified by the developer. move_room() in the game module has been changed to return a specific value when the player moves to the final room. The game struct has been updated to include a final_room field. A function add_final_room_to_game() has been written to add a final room to the game and protect against bad inputs. All relevant documentation has been duly updated. 

Although this is a small feature, I am aiming to get this approved quickly so there is some way to end games before the demo next week.

lydiafilipe commented on src/game-state/include/game.h:
```diff
@@ -110,6 +114,17 @@ int add_player_to_game(game_t *game, player_t *player);
+/* Adds the final room to the given game
+ *
+ * Parameters:
+ *  game struct
```
Note that these are pointers not just structs. It looks like this is a problem in a lot of the documentation in this file, please change all the function comments that have this.

lydiafilipe commented on src/game-state/src/game.c:
```diff
@@ -55,6 +55,22 @@ int add_room_to_game(game_t *game, room_t *room) {
+/* See game.h */
+int add_final_room_to_game(game_t *game, room_t *final_room) {
+    room_t *check;
+    HASH_FIND(hh, game->all_rooms, final_room->room_id, strnlen(final_room->room_id, MAX_ID_LEN),
```
Some of these lines are too long, please make sure they don't exceed 80 characters

lydiafilipe commented on src/game-state/src/game.c:
```diff
@@ -103,6 +119,10 @@ int move_room(game_t *game, room_t *new_room) {
         return 3;
+    if(new_room == game->final_room)
+    {
+        return 4;
```
See comment in game.h about this

lydiafilipe commented on src/game-state/include/game.h:
```diff
@@ -63,6 +66,7 @@ game_t *game_new(char *start_desc);
  *  1 for failure
  *  2 if game null
  *  3 if new_room is null
+ *  4 if new_room is the game's final_room
```
Off of Lindsey's comment, it might actually be good to have meaningful macros for all of these in addition to SUCCESS and FAILURE. I believe in the large pr I let this go but it's important to add for clarity

lydiafilipe commented:
@lindsey-hsi generally it is good to leave comments in the code directly rather than as a comment on the entire issue. It's easier to pinpoint and have direct discussion on

thoranderskow commented on docs/GAME-STATE.md:
```diff
@@ -94,14 +96,27 @@ This functions mallocs space for a new game struct. It does not take any argumen
++ Returns: 
+    - SUCCESS if successful, FAILURE if failed
+
+ Adds the final room to the given game
```
I think this comment could be more clear. Is there a way to make it clear that final room is referring to final_room within the game_t struct?

patrickkoopmans commented:
> Everything looks great for the most part! There are just a few small errors I saw, mostly in the documentation:
> 
> * move_room in game.h: the header comment reads that the function returns 0 for success and 1 for failure (alongside the other ints that specify different results), but the actual function returns 0 for FAILURE and 1 for SUCCESS, which is consistent with the conventions we’ve been following.
> * add_final_room_to_game in game.h: shouldn’t the error message read “this room already exists in the list of rooms”, since the condition is if check != NULL?
> * game-state.md: the same inconsistency applies as in the documentation for move_room in game.h — the description for move_room should read that it returns 0 for failure, 1 for SUCCESS, etc.
> 
> Other than those three things, your work looks great - it's well-written + concise, and the code and documentation follow the stylistic conventions we've been using.

Thanks for the feedback! 

1. Regarding your first point, I have changed the functions to return the correct SUCCESS and FAILURE macros
2. Actually, this check is to ensure that the room that WDL passes to load into the final_room() pointer in the game struct is an actual room in the game. This error message is to prevent WDL from making the final room a room that does not exist for the player to enter
3. Noted and changed!

lydiafilipe commented on src/game-state/src/game.c:
```diff
@@ -99,10 +120,14 @@ room_t *find_room_from_game(game_t *game, char* room_id) {
-        return 2;
+        return GAME_NULL;
     }
     if(new_room == NULL)
```
very small thing but you should have braces in this if-statement. Also, it looks like you're using k&r style but below is allman, please make sure this is consistent

lydiafilipe commented:
Issue Score: ✓

Comments:
Nice work on this pr! However there is no closing statement here.

<!-- {"score": "check"} -->
